File: src/main/java/io/github/hhy/bookmark/storage/LocalFileStorage.java
package io.github.hhy.bookmark.storage;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.io.FileUtil;
import io.github.hhy.bookmark.element.Element;
import io.github.hhy.bookmark.element.ElementBuilder;
import io.github.hhy.bookmark.element.Type;
import io.github.hhy.bookmark.util.FDUtil;
import org.apache.commons.lang.StringUtils;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static io.github.hhy.bookmark.element.Element.elementEq;
import static io.github.hhy.bookmark.util.FDUtil.PROJECT_RELATIVE;

public class LocalFileStorage implements Storage {

    public static String DEFAULT_FILE = ".idea" + File.separator + "bookmarks.json";

    private Project project;

    private final List<Element> elements = new ArrayList<>();

    private Path storeFile;

    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public LocalFileStorage(Project project) {
        this.project = project;
        this.storeFile = Path.of(project.getBasePath(), DEFAULT_FILE);
        this.checkFile();
    }

    private void checkFile() {
        File f = storeFile.toFile();
        try {
            if (!f.exists()) {
                f.createNewFile();
            }
            elements.addAll(readFromLocalFile());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<Element> getElements() throws IOException {
        return Collections.unmodifiableList(elements);
    }

    @Override
    public synchronized void addElements(List<Element> adds) throws IOException {
        List<Element> rem = new ArrayList<>();
        for (Element ele : adds) {
            List<Element> matched = elements.stream().filter(item -> elementEq(item, ele)).toList();
            rem.addAll(matched);
        }
        elements.removeAll(rem);
        elements.addAll(adds);
    }

    @Override
    public synchronized void removeElement(Element element) throws IOException {
        Optional<Element> r = elements.stream().filter(item -> elementEq(item, element)).findFirst();
        r.ifPresent(elements::remove);
    }

    @Override
    public void storage() throws IOException {
        if (elements.size() == 0) {
            Files.write(storeFile, new byte[0]);
            return;
        }
        FileUtil.writeToFile(storeFile.toFile(), gson.toJson(elements.stream().map(item -> {
            if (item.elementType() != Type.BOOKMARK) {
                return item;
            }
            if (item.fileDescriptor().startsWith(project.getBasePath())) {
                return ElementBuilder.anElement()
                        .withElementType(item.elementType())
                        .withName(item.name())
                        .withFileDescriptor(FDUtil.toRelative(item.fileDescriptor(), project.getBasePath()))
                        .withLinenumber(item.linenumber())
                        .withIndex(item.index())
                        .withBookmarkType(item.bookmarkType())
                        .withGroup(item.group())
                        .build();
            }
            return item;
        }).sorted().toList()), Charset.defaultCharset());
    }

    private List<Element> readFromLocalFile() throws IOException {
        String fileStr = Files.readString(storeFile);
        if (StringUtils.isEmpty(fileStr)) Collections.emptyList();

        return gson.fromJson(fileStr, new TypeToken<List<Element>>() {
        }).stream().map(item -> {
            if (item.elementType() != Type.BOOKMARK) {
                return item;
            }
            if (item.fileDescriptor().startsWith(PROJECT_RELATIVE)) {
                return ElementBuilder.anElement()
                        .withElementType(item.elementType())
                        .withName(item.name())
                        .withFileDescriptor(FDUtil.toAbsolute(item.fileDescriptor(), project.getBasePath()))
                        .withLinenumber(item.linenumber())
                        .withIndex(item.index())
                        .withBookmarkType(item.bookmarkType())
                        .withGroup(item.group())
                        .build();
            }
            return item;
        }).toList();
    }
}

File: src/main/java/io/github/hhy/bookmark/storage/LocalFileStorage.kt
package io.github.hhy.bookmark.storage

import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.intellij.openapi.application.TransactionGuard
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.vfs.VirtualFileManager
import io.github.hhy.bookmark.element.*
import io.github.hhy.bookmark.util.FDUtil
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

class LocalFileStorage(private val project: Project) : Storage {

    companion object {
        @JvmField
        val DEFAULT_FILE = ".idea${File.separator}bookmarks.json"
    }

    val storeFile: Path = project.basePath?.let { Path.of(it, DEFAULT_FILE) }!!
    val GSON = GsonBuilder().setPrettyPrinting().create()
    val groups: MutableMap<String, GroupElement> = HashMap()

    init {
        checkFile()
    }

    @Synchronized
    override fun elements(): List<GroupElement> = ArrayList(groups.values)

    @Synchronized
    override fun getGroup(name: String): GroupElement? = this.groups[name]

    @Synchronized
    override fun addGroup(ele: GroupElement) {
        if (ele.name !in this.groups) {
            this.groups[ele.name] = ele
        }
    }

    @Synchronized
    override fun removeGroup(name: String): GroupElement? = this.groups.remove(name)

    @Synchronized
    override fun renameGroup(groupName: String, newGroupName: String) {
        this.groups[newGroupName] = this.groups[groupName]!!
        removeGroup(groupName)
    }

    @Synchronized
    override fun getBookmark(key: String): BookmarkElement? {
        return this.groups.values.map { it.bookmarks }
            .flatten().firstOrNull { it.key() == key }
    }

    @Synchronized
    override fun addBookmark(groupName: String, ele: BookmarkElement) {
        val group: GroupElement = getGroup(groupName) ?: Element.withGroup(groupName).also { addGroup(it) }
        val bookmark = getBookmark(ele.key())
        if (bookmark == null) {
            group.bookmarks.add(0, ele)
        }
    }

    @Synchronized
    override fun removeBookmark(groupName: String, key: String): BookmarkElement? {
        val group: GroupElement = getGroup(groupName) ?: Element.withGroup(groupName).also { addGroup(it) }
        return getBookmark(key)?.also { group.bookmarks.remove(it) }
    }

    @Synchronized
    override fun storage() {
        if (groups.isEmpty()) {
            Files.writeString(storeFile, "{}")
            return
        }
        val bookmarks: Map<String, List<BookmarkElement>> = groups.mapValues { (_, group) ->
            group.bookmarks.map {
                val iit = it.clone()
                iit.fileDescriptor = FDUtil.toRelative(iit.fileDescriptor, project.basePath)
                iit
            }
        }
        FileUtil.writeToFile(
            storeFile.toFile(),
            GSON.toJson(bookmarks),
            StandardCharsets.UTF_8
        )
        TransactionGuard.getInstance().submitTransactionAndWait {
            val docManager = FileDocumentManager.getInstance()
            VirtualFileManager.getInstance().findFileByNioPath(storeFile)?.let {
                docManager.getDocument(it)
            }?.let {
                docManager.reloadFromDisk(it)
            }
        }
    }

    override fun reload() {
        TransactionGuard.getInstance().submitTransactionAndWait {
            val docManager = FileDocumentManager.getInstance()
            VirtualFileManager.getInstance().findFileByNioPath(storeFile)?.let {
                docManager.getDocument(it)
            }?.let {
                docManager.saveDocument(it)
            }
        }

        this.groups.clear()
        this.groups.putAll(readLocal())
    }

    private fun checkFile() {
        val file = storeFile.toFile()
        if (!file.exists()) {
            file.parentFile.mkdirs()
            file.createNewFile()
            Files.writeString(storeFile, "{}");
        }
    }

    @Throws(IOException::class)
    private fun readLocal(): Map<String, GroupElement> {
        val fileStr = Files.readString(storeFile)
        if (fileStr.isEmpty()) emptyList<Element>()

        val backups: Map<String, List<BookmarkElement>> = if (fileStr.startsWith("[")) {
            GSON.fromJson(fileStr, object : TypeToken<ElementList>() {})
                ?.filter { it.elementType == ElementType.BOOKMARK }
                ?.groupBy(ElementList.Item::group)?.mapValues {
                    it.value.map(ElementList.Item::toElement)
                }
                ?: return emptyMap()
        } else {
            GSON.fromJson(fileStr, object : TypeToken<Map<String, List<BookmarkElement>>>() {})
                ?: return emptyMap()
        }
        return backups.mapValues { (groupName, bookmarks) ->
            bookmarks.forEach {
                it.fileDescriptor = FDUtil.toAbsolute(it.fileDescriptor, project.basePath)
            }
            Element.withGroup(groupName, bookmarks as MutableList)
        }
    }
}


